# Pertemuan 11 - Intro to Python for Data Science
******

#  Eksplorasi dan Pengolahan Data (Pandas 1 Eksplorasi Data)
****

## Apa saja yang dipelajari hari ini ?

1. Pengenalan Pandas

2. Import Library dan Dataset

3. Membaca dan Melihat Data

4. Seleksi dan Akses Data

5. Eksplorasi Nilai Unik & Statistik

6. Analisis Awal

********

# Penjelasan Mengenai Dataset secara singkat

| Kolom         | Tipe Data       | Deskripsi                                                                     |
| ------------- | --------------- | ----------------------------------------------------------------------------- |
| `PassengerId` | Integer         | ID unik penumpang                                                             |
| `Survived`    | Kategorik (0/1) | Status selamat: `1` = selamat, `0` = tidak selamat                            |
| `Pclass`      | Kategorik (1–3) | Kelas tiket penumpang (1 = atas, 2 = tengah, 3 = bawah)                       |
| `Name`        | String          | Nama lengkap penumpang                                                        |
| `Sex`         | Kategorik       | Jenis kelamin (`male` atau `female`)                                          |
| `Age`         | Numerik         | Umur penumpang (dalam tahun), bisa kosong (`NaN`)                             |
| `SibSp`       | Integer         | Jumlah saudara/kakak/adik/istri/suami di kapal                                |
| `Parch`       | Integer         | Jumlah orang tua/anak di kapal                                                |
| `Ticket`      | String          | Nomor tiket                                                                   |
| `Fare`        | Numerik         | Harga tiket (dalam pound sterling)                                            |
| `Cabin`       | String          | Kode kabin penumpang (bisa kosong)                                            |
| `Embarked`    | Kategorik       | Pelabuhan keberangkatan: `C = Cherbourg`, `Q = Queenstown`, `S = Southampton` |


****

## Perbedaan Numpy dengan Pandas

#### 1. Pengertian Singkat
| Library    | Pengertian                                                                     |
| ---------- | ------------------------------------------------------------------------------ |
| **NumPy**  | Library Python untuk komputasi numerik dan array multidimensi.                 |
| **Pandas** | Library Python untuk manipulasi dan analisis data berbentuk tabel (dataframe). |

#### 2. Struktur Data Utama
| Library    | Struktur Data Utama             | Penjelasan Singkat                       |
| ---------- | ------------------------------- | ---------------------------------------- |
| **NumPy**  | `ndarray` (N-Dimensional Array) | Mirip array/matriks di matematika        |
| **Pandas** | `Series`, `DataFrame`           | Data tabel seperti Excel (baris & kolom) |

#### 3. Kegunaan
| Library    | Kegunaan                                                                         |
| ---------- | -------------------------------------------------------------------------------- |
| **NumPy**  | - Operasi matematika cepat (sum, mean, matrix)  <br> - Linear algebra, FFT       |
| **Pandas** | - Baca file `.csv`, `.xlsx` <br> - Filter, group, sort, missing value, statistik |

#### 4. Kapan menggunakan apa ?
| Situasi                                 | Gunakan  |
| --------------------------------------- | -------- |
| Data numerik murni (tanpa label kolom)  | ✅ NumPy  |
| Data tabular (ada nama kolom & baris)   | ✅ Pandas |
| Analisis statistik / eksplorasi dataset | ✅ Pandas |
| Perhitungan matriks / vektor            | ✅ NumPy  |

**Kesimpulan**

- NumPy = Kuat untuk angka dan array.

- Pandas = Kuat untuk tabel, cocok untuk data science dan analisis data.



## 1. Pengenalan Pandas

Pandas adalah library Open Source Python untuk analisis dan manipulasi data tabular (seperti spreadsheet atau database).
Digunakan secara luas dalam data science, machine learning, dan analisis bisnis.

Pandas adalah library Python untuk mengelola dan menganalisis data dalam bentuk tabel (seperti Excel).

> Pandas = Panel Data

### 1.1 Manfaat Pandas dalam Data Science
- Membaca, membersihkan, dan menganalisis data

- Mengolah data numerik dan kategorik

- Mengeksplorasi data

- Mempersiapkan data untuk analisis, visualisasi atau model ML (modelling)

### 1.2 Struktur Data : Series dan DataFrame
- Series: struktur data 1 dimensi, seperti kolom tunggal. Struktur data satu dimensi yang menyerupai array 1D (seperti list), namun memiliki label (index) untuk setiap elemen.

- DataFrame: struktur data 2 dimensi, seperti tabel dengan baris dan kolom.

In [ ]:
import pandas as pd

# Series
s = pd.Series([10, 20, 30, 40, 50]) # membuat data series

# DataFrame
df_example = pd.DataFrame({ #membuat data frame
    'Nama': ['Ari', 'Budi', 'Alesta', 'Wili'],
    'Umur': [25, 30, 21, 28],
    'Jurusan': ['Sistem Informasi', 'Teknik Informatika', 'Design Komunikasi Visual', 'Sastra Inggris'],
    'IPK': [3.32, 2.97, 3.3, 3.83]
})

In [3]:
s

0    10
1    20
2    30
3    40
4    50
dtype: int64

In [ ]:
df_example #pd.dataframe

Nama  Umur                   Jurusan   IPK
0     Ari    25          Sistem Informasi  3.32
1    Budi    30        Teknik Informatika  2.97
2  Alesta    21  Design Komunikasi Visual  3.30
3    Wili    28            Sastra Inggris  3.83

In [5]:
# lihat nilai series
s.values

array([10, 20, 30, 40, 50])

In [6]:
# lihat index series
s.index

RangeIndex(start=0, stop=5, step=1)

In [8]:
# akses elemen pertama series
s[3]

np.int64(40)

In [9]:
# menjumlah seluruh nilai series
s.sum()

np.int64(150)

In [10]:
# menghitung rata2 series
s.mean()

np.float64(30.0)

In [36]:
# Hitung Values DataFrame
df_example.values

array([['Ari', 25, 'Sistem Informasi', 3.32],
       ['Budi', 30, 'Teknik Informatika', 2.97],
       ['Alesta', 21, 'Design Komunikasi Visual', 3.3],
       ['Wili', 28, 'Sastra Inggris', 3.83]], dtype=object)

In [11]:
# Hitung index DataFrame
df_example.index

RangeIndex(start=0, stop=4, step=1)

In [12]:
# Menjumlahkan DataFrame
df_example.sum()

Nama                                       AriBudiAlestaWili
Umur                                                     104
Jurusan    Sistem InformasiTeknik InformatikaDesign Komun...
IPK                                                    13.42
dtype: object

In [12]:
# Menghitung Rata2 DataFrame
df_example.mean(numeric_only=True)

Umur    26.000
IPK      3.355
dtype: float64

In [14]:
df_example['Umur'].sum()

df_example['IPK'].mean()

np.float64(3.355)

In [15]:
# kalau mau bersamaan bisa pake
from IPython.display import display

print("ini menggunakan display")

display(s)
display(df_example)

print ("dibawah ini adalha menggunakan print")

print(s)
print(df_example)

ini menggunakan display


0    10
1    20
2    30
3    40
4    50
dtype: int64

Nama  Umur                   Jurusan   IPK
0     Ari    25          Sistem Informasi  3.32
1    Budi    30        Teknik Informatika  2.97
2  Alesta    21  Design Komunikasi Visual  3.30
3    Wili    28            Sastra Inggris  3.83

dibawah ini adalha menggunakan print
0    10
1    20
2    30
3    40
4    50
dtype: int64
     Nama  Umur                   Jurusan   IPK
0     Ari    25          Sistem Informasi  3.32
1    Budi    30        Teknik Informatika  2.97
2  Alesta    21  Design Komunikasi Visual  3.30
3    Wili    28            Sastra Inggris  3.83


In [43]:
print(s)
print(df_example)

0    10
1    20
2    30
3    40
4    50
dtype: int64
     Nama  Umur                   Jurusan   IPK
0     Ari    25          Sistem Informasi  3.32
1    Budi    30        Teknik Informatika  2.97
2  Alesta    21  Design Komunikasi Visual  3.30
3    Wili    28            Sastra Inggris  3.83


In [17]:
# Import library
import pandas as pd

# Contoh Series (data 1 dimensi)
s = pd.Series([10, 20, 30])
print("Contoh Series:")
display(s)

# Contoh DataFrame (data 2 dimensi)
df = pd.DataFrame({
    'Nama': ['Ari', 'Budi', 'Citra'],
    'Umur': [25, 30, 22],
    'Gaji': [5000000, 6000000, 5500000]
})
print("\nContoh DataFrame:")
display(df)


Contoh Series:


0    10
1    20
2    30
dtype: int64


Contoh DataFrame:


Nama  Umur     Gaji
0    Ari    25  5000000
1   Budi    30  6000000
2  Citra    22  5500000

| Baris Kode            | Penjelasan                                                                 |
| --------------------- | -------------------------------------------------------------------------- |
| `import pandas as pd` | Mengimpor library Pandas                                                   |
| `pd.Series(...)`      | Membuat data satu kolom (Series)                                           |
| `pd.DataFrame({...})` | Membuat tabel data dua dimensi (DataFrame)                                 |
| `display(...)`        | Menampilkan hasil lebih rapi di Jupyter Notebook (lebih baik dari `print`) |


## 2. Import Library dan Dataset
Sebelum menggunakan Pandas, kita harus meng-import library-nya terlebih dahulu ke dalam Python environment (misalnya di Jupyter Notebook). Sama seperti membuka kotak alat sebelum mulai kerja.

In [19]:
# import Pandas
import pandas as pd # import library pandas

# Membaca file CSV
df = pd.read_csv("train.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [52]:
print(df)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

In [19]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

Konvensi as pd digunakan supaya kita bisa memanggil fungsi Pandas secara singkat, misalnya pd.DataFrame().

## 3. Membaca dan Melihat Data
Melihat isi dan struktur awal dari data untuk memahami jumlah, bentuk, dan jenis data.

| Fungsi          | Keterangan                                                                |
| --------------- | ------------------------------------------------------------------------- |
| `df.head(n)`    | Melihat **n baris pertama** dari DataFrame (default 5 baris)              |
| `df.tail(n)`    | Melihat **n baris terakhir** dari DataFrame                               |
| `df.shape`      | Menampilkan jumlah **baris dan kolom**                                    |
| `df.info()`     | Menampilkan **struktur data**: jumlah non-null, tipe data tiap kolom, dll |
| `df.describe()` | Menampilkan **statistik ringkasan** (khusus untuk kolom numerik)          |
| `df.columns`    | Menampilkan daftar **nama kolom**                                         |
| `df.index`      | Menampilkan daftar **index**                                              |
| `df.dtypes`     | Menampilkan **tipe data** dari tiap kolom                                 |


In [21]:
df.head()        # 5 baris pertama atau melihat data dari atas

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [23]:
df.tail(10)        # 5 baris terakhir atau melihat data dari bawah

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

In [24]:
df.shape         # (jumlah baris, jumlah kolom)

(891, 12)

In [23]:
df.info()        # Struktur kolom, jumlah data non-null, tipe data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [24]:
df.describe()    # Statistik ringkasan untuk data numerik

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

| Nama Kolom | Arti                                                     |
| ---------- | -------------------------------------------------------- |
| `count`    | Jumlah data yang **tidak kosong (non-null)**             |
| `mean`     | Rata-rata (nilai tengah)                                 |
| `std`      | Standar deviasi (seberapa tersebar nilai dari rata-rata) |
| `min`      | Nilai minimum                                            |
| `25%`      | Kuartil pertama (25% dari data < nilai ini)              |
| `50%`      | Median atau kuartil kedua (nilai tengah data)            |
| `75%`      | Kuartil ketiga (75% dari data < nilai ini)               |
| `max`      | Nilai maksimum                                           |


In [25]:
df.columns       # Nama-nama kolom

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [26]:
df.index         # Informasi indeks baris

RangeIndex(start=0, stop=891, step=1)

In [27]:
df.dtypes        # Tipe data masing-masing kolom

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## 4. Seleksi dan Akses Data
Kita bisa mengakses data tertentu berdasarkan kolom atau baris.

### Contoh DataFrame

In [27]:
import pandas as pd

df = pd.DataFrame({
    'Nama': ['Ari', 'Budi', 'Cici'],
    'Umur': [25, 30, 22]
}, index=['a', 'b', 'c'])

df

Nama  Umur
a   Ari    25
b  Budi    30
c  Cici    22

#### df.loc[] → berdasarkan label index

In [28]:
df.loc['c'] # Akses baris dengan label 'a'

Nama    Cici
Umur      22
Name: c, dtype: object

In [30]:
df.loc['a', 'Nama']  # Akses nama orang di baris 'b'

'Ari'

In [32]:
df.loc[:, 'Umur']    # Semua baris, kolom 'Umur'

a    25
b    30
c    22
Name: Umur, dtype: int64

#### df.iloc[] → berdasarkan posisi baris

In [34]:
df.iloc[0]      # Akses baris pertama (posisi 0)

Nama    Ari
Umur     25
Name: a, dtype: object

In [ ]:
# 1 = nyari baris, 0 = nyari kolom
df.iloc[1, 0]   # Akses baris ke-2, kolom ke-1 (Budi)

'Budi'

In [37]:
df.iloc[:, 0]    # Semua baris, kolom ke-2 (Umur)

a     Ari
b    Budi
c    Cici
Name: Nama, dtype: object

**df.loc[0] akan error kalau index-nya bukan angka 0.**

**df.iloc['a'] akan error karena 'a' bukan angka.**

```lua
        loc
         ↓
+-----+--------+------+
| a   | Ari    |  25  |
| b   | Budi   |  30  |
| c   | Cici   |  22  |
+-----+--------+------+
```
```lua
        iloc
         ↓
+-----+--------+------+
|  0  | Ari    |  25  |
|  1  | Budi   |  30  |
|  2  | Cici   |  22  |
+-----+--------+------+
```

| Fungsi    | Mengakses berdasarkan    | Cocok untuk             |
| --------- | ------------------------ | ----------------------- |
| `df.loc`  | Label index & nama kolom | DataFrame dengan label  |
| `df.iloc` | Posisi numerik (0,1,2,…) | Posisi absolut / urutan |


### Berdasarkan Dataset CSV

In [38]:
import pandas as pd

df = pd.read_csv('train.csv')

In [39]:
# Mengakses kolom nama
# df['Name']
df[['Name']]

Name
0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
..                                                 ...
886                              Montvila, Rev. Juozas
887                       Graham, Miss. Margaret Edith
888           Johnston, Miss. Catherine Helen "Carrie"
889                              Behr, Mr. Karl Howell
890                                Dooley, Mr. Patrick

[891 rows x 1 columns]

In [40]:
# Mengakses nama, age dan gender
df[['Name', 'Age', 'Sex']]

Name   Age     Sex
0                              Braund, Mr. Owen Harris  22.0    male
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0  female
2                               Heikkinen, Miss. Laina  26.0  female
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0  female
4                             Allen, Mr. William Henry  35.0    male
..                                                 ...   ...     ...
886                              Montvila, Rev. Juozas  27.0    male
887                       Graham, Miss. Margaret Edith  19.0  female
888           Johnston, Miss. Catherine Helen "Carrie"   NaN  female
889                              Behr, Mr. Karl Howell  26.0    male
890                                Dooley, Mr. Patrick  32.0    male

[891 rows x 3 columns]

In [41]:
# Mengakses Umur aja
df[['Age']]

Age
0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
..    ...
886  27.0
887  19.0
888   NaN
889  26.0
890  32.0

[891 rows x 1 columns]

In [67]:
# Akses umur
df.Age

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [68]:
# Ambil beberapa kolom
df[['Name', 'Age', 'Survived']]

Name   Age  Survived
0                              Braund, Mr. Owen Harris  22.0         0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0         1
2                               Heikkinen, Miss. Laina  26.0         1
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0         1
4                             Allen, Mr. William Henry  35.0         0
..                                                 ...   ...       ...
886                              Montvila, Rev. Juozas  27.0         0
887                       Graham, Miss. Margaret Edith  19.0         1
888           Johnston, Miss. Catherine Helen "Carrie"   NaN         0
889                              Behr, Mr. Karl Howell  26.0         1
890                                Dooley, Mr. Patrick  32.0         0

[891 rows x 3 columns]

In [43]:
# Penumpang yang selamat
df[df['Survived'] == 0]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
6              7         0       1                   McCarthy, Mr. Timothy J   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch           Ticket     Fare Cabin Embarked  
0      male  22.0      1      0        A/5 21171   7.2500   NaN        S  
4      male  35.0      0      0           373450   8.0500   NaN        S  
5      male   NaN      0      0           330877   8.4583   NaN        Q  
6      male  54.0      0      0            17463  51.8625   E46        S  
7      male   2.0      3      1           349909  21.0750   NaN        S  
..      ...   ...    ...    ...              ...      ...   ...      ...  
884    male  25.0      0      0  SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5           382652  29.1250   NaN        Q  
886    male  27.0      0      0           211536  13.0000   NaN        S  
888  female   NaN      1      2       W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0           370376   7.7500   NaN        Q  

[549 rows x 12 columns]

In [45]:
# Mengakses baris pertama berdasarkan label
df.loc[777]                # Baris pertama berdasarkan label

PassengerId                              778
Survived                                   1
Pclass                                     3
Name           Emanuel, Miss. Virginia Ethel
Sex                                   female
Age                                      5.0
SibSp                                      0
Parch                                      0
Ticket                                364516
Fare                                  12.475
Cabin                                    NaN
Embarked                                   S
Name: 777, dtype: object

In [46]:
# Mengakses baris dan kolom tertentu
df.loc[10, 'Name']

'Sandstrom, Miss. Marguerite Rut'

In [47]:
# Filter Baris Berdasarkan Kondisi
# Lihat nama dan umur penumpang pertama
df.loc[777, ['Name', 'Age', 'Fare', 'Embarked']]

Name        Emanuel, Miss. Virginia Ethel
Age                                   5.0
Fare                               12.475
Embarked                                S
Name: 777, dtype: object

In [51]:
# Ambil nama dan umur penumpang pertama sebagai DataFrame
pd.DataFrame(df.loc[[0], ['Name', 'Age', 'Fare', 'Cabin']])

Name   Age  Fare Cabin
0  Braund, Mr. Owen Harris  22.0  7.25   NaN

In [54]:
df.iloc[50]               # Baris pertama berdasarkan posisi

PassengerId                            51
Survived                                0
Pclass                                  3
Name           Panula, Master. Juha Niilo
Sex                                  male
Age                                   7.0
SibSp                                   4
Parch                                   1
Ticket                            3101295
Fare                              39.6875
Cabin                                 NaN
Embarked                                S
Name: 50, dtype: object

In [ ]:
# baris dan kolom
df.iloc[0, 3]

'Braund, Mr. Owen Harris'

## 5. Eksplorasi Nilai Unik & Statistik
Melihat distribusi data kategorikal dan berapa banyak nilai unik yang dimiliki suatu kolom.

**Dalam eksplorasi data menggunakan Pandas, kita sering ingin tahu:**

- Berapa nilai yang berbeda (unik) ada dalam suatu kolom?

- Apa saja nilai yang muncul di kolom tersebut?

- Seberapa sering setiap nilai muncul?

### 5.1 .value_counts()
Menghitung jumlah kemunculan setiap nilai unik dalam suatu kolom.

In [55]:
import pandas as pd

df = pd.read_csv('train.csv')

In [56]:
df['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [100]:
# Hitung berapa banyak penumpang yang selamat dan tidak selamat
print(df['Survived'].value_counts())

Survived
0    549
1    342
Name: count, dtype: int64


### 5.2 .unique()

Menampilkan daftar nilai unik (tanpa duplikasi) dari suatu kolom.

In [57]:
print(df['Pclass'].unique())
# Artinya: Terdapat 3 kelas penumpang: kelas 3, kelas 1, dan kelas 2.

[3 1 2]


In [192]:
df['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [58]:
df['Sex'].unique()        # Menampilkan nilai unik: ['male', 'female']

array(['male', 'female'], dtype=object)

### 5.3 .nunique()

In [59]:
df['Pclass'].nunique()       # Jumlah nilai unik: 2

3

In [194]:
df['Pclass'].nunique()

3

In [60]:
# Contoh Gabungan
# Berapa jenis gender yang ada?
print("Unique genders:", df['Sex'].unique())
print("Jumlah gender unik:", df['Sex'].nunique())
print("Jumlah masing-masing gender:")
print(df['Sex'].value_counts())


Unique genders: ['male' 'female']
Jumlah gender unik: 2
Jumlah masing-masing gender:
Sex
male      577
female    314
Name: count, dtype: int64


| Tujuan                                         | Fungsi Pandas     |
| ---------------------------------------------- | ----------------- |
| Menampilkan apa saja nilainya (tanpa duplikat) | `.unique()`       |
| Menghitung berapa jenis nilai yang ada         | `.nunique()`      |
| Mengetahui jumlah kemunculan tiap nilai        | `.value_counts()` |


## 6. Analisis Awal
### Tujuan:
- Mengetahui apakah data ada yang hilang (missing values).

- Mengetahui hubungan (korelasi) antar fitur numerik.
### 6.1  Mendeteksi missing values: isnull().sum()
- Menampilkan jumlah data kosong (NaN) pada setiap kolom. Penting untuk membersihkan data.

- Digunakan untuk mencari apakah ada data yang kosong/null di dataset.

- Mengetahui kolom mana yang memiliki data kosong (NaN).

- Membantu memutuskan strategi pembersihan data: apakah ingin di-drop, diisi dengan rata-rata, median, modus, atau lainnya.

- Meningkatkan kualitas data sebelum dilakukan analisis atau modeling.

**Bayangin kamu punya daftar temen yang RSVP buat dateng ke pesta kamu. Tapi ternyata, ada yang nggak ngisi nama atau alamat—nah isnull().sum() itu kayak kamu ngecek,**

**“Eh, siapa aja nih yang datanya belum lengkap?”**

In [61]:
import pandas as pd

df = pd.read_csv('train.csv')

# Cek jumlah missing values di tiap kolom
print(df.isnull().sum())

# Kolom Age memiliki 177 data kosong
# Kolom Cabin memiliki 687 data kosong
# Embarked hanya 2 data kosong

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [107]:
# Mengecek jumlah missing values di setiap kolom
missing_values = df.isnull().sum()

# Tampilkan hanya kolom yang punya missing values
print(missing_values[missing_values > 0])

# Kolom Age memiliki 177 baris kosong
# Cabin bahkan lebih dari setengah data kosong (bisa dipertimbangkan untuk dibuang)
# Embarked hanya 2 data kosong, bisa diisi manual

Age         177
Cabin       687
Embarked      2
dtype: int64


### 6.2 Korelasi antar kolom: corr()
- Melihat hubungan antar kolom numerik. Nilai mendekati 1 atau -1 menunjukkan hubungan kuat.

- Korelasi digunakan untuk melihat seberapa kuat hubungan antar kolom numerik.

- Mengetahui hubungan antar variabel numerik (misalnya antara Fare dan Survived).

- Membantu memilih fitur yang relevan untuk model machine learning (feature selection).

- Mendeteksi multikolinearitas (fitur yang saling berkorelasi tinggi dan bisa mengganggu model prediktif).

In [62]:
df.corr(numeric_only=True)

PassengerId  Survived    Pclass       Age     SibSp     Parch  \
PassengerId     1.000000 -0.005007 -0.035144  0.036847 -0.057527 -0.001652   
Survived       -0.005007  1.000000 -0.338481 -0.077221 -0.035322  0.081629   
Pclass         -0.035144 -0.338481  1.000000 -0.369226  0.083081  0.018443   
Age             0.036847 -0.077221 -0.369226  1.000000 -0.308247 -0.189119   
SibSp          -0.057527 -0.035322  0.083081 -0.308247  1.000000  0.414838   
Parch          -0.001652  0.081629  0.018443 -0.189119  0.414838  1.000000   
Fare            0.012658  0.257307 -0.549500  0.096067  0.159651  0.216225   

                 Fare  
PassengerId  0.012658  
Survived     0.257307  
Pclass      -0.549500  
Age          0.096067  
SibSp        0.159651  
Parch        0.216225  
Fare         1.000000

| Kolom yang Dikorelasikan | Nilai Korelasi | Artinya (Bahasa Gaul)                                                                                   |
| ------------------------ | -------------- | ------------------------------------------------------------------------------------------------------- |
| `Survived` vs `Pclass`   | -0.33          | Semakin mahal kelas (1), makin besar kemungkinan selamat. Kelas 3 banyak yang ngga selamat.             |
| `Survived` vs `Fare`     | **0.25**       | Semakin mahal bayar tiket, makin besar peluang selamat. Mungkin karena orang kaya duluan diselamatin 😬 |
| `Survived` vs `Age`      | -0.07          | Umur gak terlalu ngaruh ke selamat atau engga. Korelasi lemah banget.                                   |
| `Survived` vs `SibSp`    | -0.03          | Jumlah saudara/istri/suami gak terlalu ngaruh juga.                                                     |
| `Survived` vs `Parch`    | 0.08           | Ada sedikit pengaruh kalau dia bawa ortu/anak, tapi masih lemah korelasinya.                            |


- Orang yang naik kelas 1 atau bayar tiket mahal (Fare tinggi) punya kemungkinan lebih besar untuk selamat.

- Umur, jumlah keluarga ternyata gak terlalu ngaruh, cuma korelasinya dikit banget.

- Pclass dan Fare itu kebalikan banget (lihat -0.54), wajar karena kelas 1 pasti bayarnya mahal.

In [109]:
df[['Fare', 'Survived']].corr()

Fare  Survived
Fare      1.000000  0.257307
Survived  0.257307  1.000000

Artinya : beri tau saya seberapa nyambung antara orang yang bayar mahal (Fare) sama kemungkinan dia selamat (Survived)".

|              | Fare     | Survived |
| ------------ | -------- | -------- |
| **Fare**     | 1.000000 | 0.257307 |
| **Survived** | 0.257307 | 1.000000 |

Arti Angkanya:
- 1.0 = korelasi dirinya sendiri (ya pasti 100% cocok lah)

- 0.257307 = ini yang penting:

Ada korelasi positif lemah antara Fare dan Survived.

Maksudnya: Semakin mahal tiket yang dibayar, kemungkinan dia selamat lebih tinggi.

### Membaca Korelasi:
- Nilai mendekati +1 → hubungan kuat positif

- Nilai mendekati -1 → hubungan kuat negatif

- Nilai mendekati 0 → tidak ada hubungan atau sangat lemah

### Perbedaan dan Kapan digunakan

| Analisis                    | Fungsi           | Kapan Digunakan                                                                      | Tujuan                                                                            | Output                            |
| --------------------------- | ---------------- | ------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------- | --------------------------------- |
| ✅ Mendeteksi Missing Values | `isnull().sum()` | **Tahap awal eksplorasi data** (sebelum preprocessing atau modeling)                 | Mengetahui kolom mana yang memiliki data hilang/null dan berapa jumlahnya         | Jumlah nilai kosong per kolom     |
| 🔗 Korelasi antar kolom     | `corr()`         | **Setelah tahu data numerik bersih**, atau sebelum memilih fitur (feature selection) | Mengetahui **hubungan antar fitur numerik**: apakah saling berkorelasi atau tidak | Matriks korelasi (angka -1 s/d 1) |


### Kesimpulan :
| Fungsi           | Tujuan Utama                                  |
| ---------------- | --------------------------------------------- |
| `isnull().sum()` | Cek kebersihan data (Missing Values)          |
| `corr()`         | Lihat hubungan antar fitur numerik (Korelasi) |

Keduanya penting dan berada di tahap awal eksplorasi data, namun digunakan untuk tujuan yang berbeda.

***************

## Latihan

1. Tampilkan 10 data pertama dari dataset Titanic.

2. Berapa jumlah penumpang laki-laki dan perempuan?

3. Hitung rata-rata umur penumpang.

4. Kolom apa saja yang memiliki missing value?

5. Berapa nilai korelasi antara Fare dan Survived?

In [ ]:
# nomor 1
import pandas as pd

# Membaca file CSV
df = pd.read_csv('train.csv')

#Tampilkan 10 data pertama
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [ ]:
# 2. Jumlah penumpang pria dan wanita
df['Sex'].value_counts()


Sex
male      577
female    314
Name: count, dtype: int64

In [ ]:
# 3. Rata-rata umur penumpang
df['Age'].mean()

np.float64(29.69911764705882)

In [4]:
# 4. Kolom dengan missing value
df.isnull().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [5]:
# 5. Korelasi antara Fare dan Survived
df[['Fare', 'Survived']].corr()

Fare  Survived
Fare      1.000000  0.257307
Survived  0.257307  1.000000